### PR TITLE
lib/vfscore: Auto-mounting scheme with 3 tables

### PR DESCRIPTION
**Pull request by skuenzer**
### Description of changes

This PR introduces a fairly flexible mounting mechanism by
using three filesystem tables:
1) compiled-in, configured via menuconfig
2) user-provided, read entries from kernel parameters ("vfs.fstab")
3) fall-back, similar to "compiled-in"

If auto-mounting is activated, entries of the comiled-in table (1) are mounted first. Afterwards, compiled-in fall-back entries (2) are only processed if the user did not provide any mount entries via the kernel command line (3). The compiled-in tables can be empty and can keep up to four entries.

As part of this PR, The `extract` option of `initrd` replaces the virtual `initrd` filesystem driver. This is done in order to disconnect archive extraction from filesystem mounting and provides greater flexibility for setting up a VFS layout. The traditional initrd mode can be achieved with two entries: First, a RAMFS volume needs to be mounted to `/`. The second entry is used to `extract` virtual fs driver to unpack a CPIO archive to the root volume `/`.

> [!CAUTION]
> This is a breaking change and this PR should be checked in conjunction with [unikraft/kraftkit#1154][0] and merged simultaneously with [#1154][0] to avoid incompatibility issues. Simultaneously, all [catalog](https://github.com/unikraft/catalog) entries should be retriggered against Unikraft's `staging` branch following the merge. Since Unikraft is pre-v1.0, no backwards compatibility is included in this refactor.

[0]: https://github.com/unikraft/unikraft/pull/1154

### CODE REVIEW

**Comment by StefanJum** [comment]
@skuenzer just to make sure I'm testing it right, selecting `User-provided filesystem table` -> `Fall-back entries` -> `9pfs` in the config and passing no kernel parameters at runtime should be equivalent to the old vfscore automount feature, right?

**Comment by skuenzer** [comment]
> @skuenzer just to make sure I'm testing it right, selecting `User-provided filesystem table` -> `Fall-back entries` -> `9pfs` in the config and passing no kernel parameters at runtime should be equivalent to the old vfscore automount feature, right?

The full equivalent is configuring 9pfs in the compiled-in table. Then this will be mounted always, irrespective of what the user provides on the kernel command line. The fall-back table is only taken into account if there is nothing provided by the user on the kernel command line.

**Review by nderjung — APPROVED** [review]
Adding my review, I have tested this and it works as expected for me. Thanks!

Reviewed-by: Alexander Jung <alex@unikraft.io>

File: lib/vfscore/automount.c [hunk 1]
```diff
@@ -50,18 +50,20 @@
...
-#ifdef CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS
 #include <errno.h>
 #include <uk/config.h>
 #include <uk/arch/types.h>
-#endif /* CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS */
 
-#if CONFIG_LIBVFSCORE_FSTAB
+#if CONFIG_LIBVFSCORE_AUTOMOUNT_UP
 #define LIBVFSCORE_FSTAB_VOLUME_ARGS_SEP			':'
 #define LIBVFSCORE_FSTAB_UKOPTS_ARGS_SEP			','
```
Comment (StefanJum) [inline lib/vfscore/automount.c]: If we do this, compiling with `Compiled-in filesystem table` and without `User-provided filesystem table` config options will fail with `vfscore/automount.c:466:41: error: LIBVFSCORE_FSTAB_UKOPTS_ARGS_SEP’ undeclared`.
Is there any reason to add these in a guard and not just always define them?

**Review by StefanJum — CHANGES_REQUESTED** [review]
See the one comment, all good besides that.

**Review by razvand — APPROVED** [review]
Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

File: lib/vfscore/automount.c (hunk 1 above)
Comment (skuenzer) [reply lib/vfscore/automount.c]: Ouh, good point, there is one tiny mistake then. But you have actually good point, they don't harm if they are unused, no memory will be used. I'll remove this guard. Thanks for spotting this.

**Comment by skuenzer** [comment]
> Reviewed-by: Razvan Deaconescu [razvand@unikraft.io](mailto:razvand@unikraft.io) Approved-by: Razvan Deaconescu [razvand@unikraft.io](mailto:razvand@unikraft.io)

Please note that I added another commit on top because I noticed that ukopts where implicitly deconst'd while parsing. There is a bit more described in the commit subject, however, this change has effect of configuring Unikraft mount options, so I assume it affects maybe KraftKit. please have a look @razvand, @nderjung .

**Review by StefanJum — APPROVED** [review]
All good on my side now, thanks.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

**Review by nderjung — APPROVED** [review]
> Please note that I added another commit on top because I noticed that ukopts where implicitly deconst'd while parsing. There is a bit more described in the commit subject, however, this change has effect of configuring Unikraft mount options, so I assume it affects maybe KraftKit. please have a look @razvand, @nderjung .

I have re-tested the changes with [the accompanying changes to KraftKit](https://github.com/unikraft/kraftkit/pull/1154) and everything still works as expected.

Reviewed-by: Alexander Jung <alex@unikraft.io>

**Comment by razvand** [comment]
I'm testing it myself before the final approval.

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request merged**